lorawan-encoding/src/certification.rs

```diff
@@ -0,0 +1,136 @@
+    #[cmd(cid = 0x04, len = 1)]
+    AdrBitChangeReq(AdrBitChangeReqPayload<'a>),
+
+    /// Change uplink periodicity to the provided value
+    #[cmd(cid = 0x06, len = 1)]
+    TxPeriodicityChangeReq(TxPeriodicityChangeReqPayload<'a>),
+
+    /// Send all subsequent uplinks of the specified type
+    // NB! Variable length payload
+    #[cmd(cid = 0x07)]
+    TxFramesCtrlReq(TxFramesCtrlReqPayload<'a>),
+
+    /// Request to send firmware version, LoRaWAN version, and regional parameters version
+    #[cmd(cid = 0x7f, len = 0)]
+    DutVersionsReq(DutVersionsReqPayload),
+}
+
+#[derive(Debug, PartialEq, CommandHandler)]
+#[cfg_attr(feature = "defmt-03", derive(defmt::Format))]
+pub enum UplinkDUTCommand<'a> {
+    #[cmd(cid = 0x7f, len = 12)]
+    DutVersionsAns(DutVersionsAnsPayload<'a>),
+}
+
+pub fn parse_downlink_certification_messages(
+    data: &[u8],
+) -> MacCommandIterator<'_, DownlinkDUTCommand<'_>> {
+    MacCommandIterator::new(data)
+}
+
+impl AdrBitChangeReqPayload<'_> {
+    /// Enable/disable ADR
+    pub fn adr_enable(&self) -> Result<bool, Error> {
+        match self.0[0] {
+            0 => Ok(false),
+            1 => Ok(true),
+            _ => Err(Error::RFU),
+        }
+    }
+}
+
+impl DutVersionsAnsCreator {
+    pub fn set_versions_raw(&mut self, data: [u8; 12]) -> &mut Self {
+        self.data[0..12].copy_from_slice(&data);
+        self
+    }
+}
+
+impl TxPeriodicityChangeReqPayload<'_> {
+    pub fn periodicity(&self) -> Result<Option<u16>, Error> {
+        let v = self.0[0];
+        if v > 10 {
+            Err(Error::RFU)
+        } else if v == 0 {
+            // DUT should switch back to default behaviour
+            Ok(None)
+        } else {
+            let seconds = match v {
+                1 => 5,
+                2 => 10,
+                3 => 20,
+                4 => 30,
+                5 => 40,
+                6 => 50,
+                7 => 60,
+                8 => 120,
+                9 => 240,
+                10 => 480,
+                0_u8 | 11_u8..=u8::MAX => unreachable!(),
+            };
+            Ok(Some(seconds))
+        }
+    }
+}
+
+impl<'a> TxFramesCtrlReqPayload<'a> {
+    pub fn new(data: &'a [u8]) -> Result<Self, Error> {
+        if data.is_empty() {
+            return Err(Error::BufferTooShort);
+        }
+        Ok(TxFramesCtrlReqPayload(data))
+    }
+
+    /// Maximum possible length of the payload
+    const fn max_len() -> usize {
+        // NB! Payload might include extra 0..N octets, but
+        // TS009-1.2.1 does not specify upper limit.
+        242
+    }
+
+    const fn min_len() -> usize {
+        // Minimum length of the of payload including CID
+        2
+    }
+
+    /// Actual length of the payload
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        // Payload should have at least minimum length...
+        core::cmp::max(Self::min_len(), self.0.len())
```

Comment:
I was concerned you made a mistake here but it looks like these commands are sent only one at a time! Interesting

Reply:
Yeah, this is annoying. Probably warrants a better comment and I'll also send a note about improving the protocol documentation.